### PR TITLE
New version of kpt cfg set

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent 

This is the initial implementation of `kpt cfg set`. It uses most of the code from downstream kyaml. Currently it works for setters only. Substitution patterns will be supported in following PRs. This code is not called by kpt, so can be safely merged in to kpt master instead of feature branch. Will add more tests in coming PRs. Tried to make the PR short and main aim here is to get the approval on overall direction.

### CODE REVIEW

File: internal/cmdset/cmdset.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		IsSet:              true,
+	}
+	count, err := r.Set.Set()
+	if err != nil {
+		// return err if RecurseSubPackages is false
+		if !r.Set.RecurseSubPackages {
+			return err
+		}
+		// print error message and continue if RecurseSubPackages is true
+		fmt.Fprintf(w, "%s\n", err.Error())
```
Comment (mortent) [inline internal/cmdset/cmdset.go:120]: Should we try to separate executing the command from printing? It doesn't matter that much in this example, but this is an area where I think we want to do better than the existing commands.

File: internal/cmdset/cmdset.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		// else it is a scalar value
+		r.Value = setterValue
+	}
+	return nil
+}
+
+func (r *SetRunner) runE(c *cobra.Command, args []string) error {
+	e := runner.ExecuteCmdOnPkgs{
+		NeedOpenAPI:        true,
+		Writer:             c.OutOrStdout(),
```
Comment (mortent) [inline internal/cmdset/cmdset.go:87]: Didn't we have a discussion a while back about removing this property?

File: internal/cmdset/cmdset.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+func (r *SetRunner) preRunE(c *cobra.Command, args []string) error {
+	valueFlagSet := c.Flag("value").Changed
+
+	if !valueFlagSet {
+		return errors.Errorf("value flag must be set")
+	}
+
+	// r.KeyValue holds SETTER_NAME=SETTER_VALUE, split them into name and value
+	r.Name = strings.Split(r.KeyValue, "=")[0]
```
Comment (mortent) [inline internal/cmdset/cmdset.go:66]: Do we need to split it twice? And we should check the size of the slice before accessing the elements.

Comment (mortent) [reply internal/cmdset/cmdset.go:66]: Also, I seem to remember discussions around supporting multiple key-value pairs?

File: internal/cmdset/cmdset_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/kustomize/kyaml/fieldmeta"
+	"sigs.k8s.io/kustomize/kyaml/openapi"
+)
+
+func TestSetCommand(t *testing.T) {
```
Comment (mortent) [inline internal/cmdset/cmdset_test.go:16]: We should add tests for how we handle incorrect input.

File: internal/cmdset/cmdset_test.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  - "11"
+ `,
+		},
+	}
+	for i := range tests {
+		test := tests[i]
+		t.Run(test.name, func(t *testing.T) {
+			// reset the openAPI afterward
+			openapi.ResetOpenAPI()
+			defer openapi.ResetOpenAPI()
```
Comment (mortent) [inline internal/cmdset/cmdset_test.go:127]: Is this still needed? Are we still mutating the openapi schema?

File: internal/util/setters/fieldmeta.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+// processShortHand parses the comment for short hand ref, loads schema to fm
+// and returns true if successful, returns false for any other cases and not throw
+// error, as the comment might not be a setter ref
+func (fm *FieldMeta) processShortHand(comment string) bool {
+	input := map[string]string{}
+	err := json.Unmarshal([]byte(comment), &input)
+	if err != nil {
+		return false
+	}
+	name := input[shortHandRef]
```
Comment (mortent) [inline internal/util/setters/fieldmeta.go:81]: Do we need to reference a `shortHandRef` here? Do we support anything else?

File: internal/util/setters/fieldmeta.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	return false
+}
+
+// FieldValueType defines the type of input to register
+type FieldValueType string
+
+const (
+	// CLIDefinitionsPrefix is the prefix for cli definition keys.
+	CLIDefinitionsPrefix = "io.k8s.cli."
```
Comment (mortent) [inline internal/util/setters/fieldmeta.go:126]: Do we still use these prefixes?

File: internal/cmdset/cmdset.go (hunk 1 above)
Comment (phanimarupaka) [reply internal/cmdset/cmdset.go:120]: ExecuteCmd method at cli layer will only print the out message here. The libraries won't print any messages. I think this is simple and good enough for commands like this. What do you mean by separating further ? Wrap it in a different method?

File: internal/cmdset/cmdset.go (hunk 2 above)
Comment (phanimarupaka) [reply internal/cmdset/cmdset.go:87]: Done. imported runner code as well.

File: internal/cmdset/cmdset.go (hunk 3 above)
Comment (phanimarupaka) [reply internal/cmdset/cmdset.go:66]: Done. 

Yes. I want to start off simple and expand it further.

File: internal/cmdset/cmdset_test.go (hunk 4 above)
Comment (phanimarupaka) [reply internal/cmdset/cmdset_test.go:16]: Done. More tests will be added in coming PRs.

File: internal/cmdset/cmdset_test.go (hunk 5 above)
Comment (phanimarupaka) [reply internal/cmdset/cmdset_test.go:127]: Not needed. Removed.

File: internal/util/setters/fieldmeta.go (hunk 6 above)
Comment (phanimarupaka) [reply internal/util/setters/fieldmeta.go:81]: Renamed it to `setterRef`

File: internal/util/setters/fieldmeta.go (hunk 7 above)
Comment (phanimarupaka) [reply internal/util/setters/fieldmeta.go:126]: For now yes. This PR still assumes old openAPI format for Setter Definitions. Once we finalize new version of Setter Definitions design, I will modify the openAPI part.

File: internal/cmdset/cmdset.go (hunk 1 above)
Comment (mortent) [reply internal/cmdset/cmdset.go]: This solution is certainly much better than printing within the library. It still requires us to print for each subpackage individually. It is a step in the right direction.

**Pull request closed without merging**